### DISCUSSION TITLE
getPrice() from CartManager returning empty object

### DISCUSSION THREAD

**Discussion by jonathanelmgren** [pimcore/pimcore#12194] [Q&A / Support]
Hello!

I am looking into the ecommerce bundle and trying to get a cart working. 

This is my function to get all items in the cart:

```
   public function getCartItems()
    {
        $manager = Factory::getInstance()->getCartManager();

        //Create cart if it doesn't exist
        $cart = $manager->getCartByName('cart');
        if (!$cart) {
            $cartId = $manager->createCart(['name' => 'cart']);
            $cart = $manager->getCart($cartId);
        }

        $products = [];

        foreach ($cart->getItems() as $cartItem) {
            $object = (object) [
                'id' => $cartItem->getProduct()->getId(),
                'name' => $cartItem->getProduct()->getName(),
                'price' => $cartItem->getPrice(),
                'qty' => $cartItem->getCount(),
            ];

            array_push($products, $object);
        }

        return $products;
    }
}
```

I also implemented Price Calculator and I can get the total sub sum of the cart without problems with `$cart->getPriceCalculator()->getSubTotal();`

Can anyone point me in the right direction?

**Comment by fashxp** [comment]
I'm not sure I understand the actual problem you are having?

**Reply by kingjia90** [reply]
Shouldn't it be like `$cartItem->getProduct()->getPrice()`?

**Reply by jonathanelmgren** [reply]
> I'm not sure I understand the actual problem you are having?

Sorry if I'm not clear @fashxp. I have a very limited knowledge about PHP. Coming from JS-frameworks but now having to develop an ecommerce site in Pimcore so I'm having some struggles.

In the document about Cart Manager (https://pimcore.com/docs/pimcore/current/Development_Documentation/E-Commerce_Framework/Cart_Manager.html):

<img width="462" alt="Skärmavbild 2022-05-24 kl  11 44 43" src="https://user-images.githubusercontent.com/81702447/170002691-4a14da31-5cfd-4751-8d99-1bf2c71fa09e.png">

Focusing on `$price = $item->getPrice(); `

the `$item->getPrice(); ` function returns an empty object and not a price

**Reply by jonathanelmgren** [reply]
> `$cartItem->getProduct()->getPrice()` works, but I got a feeling I should get the cart items price and not the product price per se. Thinking about priceInfo and modifications.

**Reply by kingjia90** [reply]
then probably you need `$cartItem->getProduct()->getOSPrice()`

**Reply by fashxp** [reply]
`$cartItem->getProduct()->getPrice()` is not part of the framework ... this might work by coincidence because your datamodel contains an attribute named `price`. 

did you implement and configure a price system? it seems that this is missing somehow. details see https://pimcore.com/docs/pimcore/current/Development_Documentation/E-Commerce_Framework/Working_with_Prices/Price_Systems_and_getting_Prices.html

interesting is though, that the cart price calculator works ... probably you have overwritten the corresponding functions?

**Reply by jonathanelmgren** [reply]
@fashxp I did configure a price system. I can get price with `product.OSPrice` and the price calculator works.

I haven't overwritten (as I know of) any functions but the `getPriceSystemName` and `getCategories` functions


This is my Product class

```
class Product extends \Pimcore\Bundle\EcommerceFrameworkBundle\Model\AbstractProduct implements \Pimcore\Bundle\EcommerceFrameworkBundle\Model\CheckoutableInterface
{
    public function getPriceSystemName(): ?string
    {
        return 'default';
    }

    public function getCategories(): ?array
    {
        return [];
    }
}
```

This is my config.yaml:
```
pimcore_ecommerce_framework:
    price_systems:
        default:
            id: Pimcore\Bundle\EcommerceFrameworkBundle\PriceSystem\AttributePriceSystem
```

As stated in docs I shouldn't need to configure in service.yaml when using `AttributePriceSystem`

I am lost at this point

**Reply by fashxp** [reply]
if `product.getOSPrice()` works, but `$item->getPrice()` doesn't, it is really odd ... as these two functions do pretty much the same, see [here](https://github.com/pimcore/pimcore/blob/10.x/bundles/EcommerceFrameworkBundle/CartManager/AbstractCartItem.php#L246). 

you probably have to debug that a bit on your end. don't have an idea right now either...
